core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -911,14 +916,14 @@ private[spark] class TaskSetManager(
     logDebug("Checking for speculative tasks: minFinished = " + minFinishedForSpeculation)
     if (tasksSuccessful >= minFinishedForSpeculation && tasksSuccessful > 0) {
       val time = clock.getTimeMillis()
-      val durations = taskInfos.values.filter(_.successful).map(_.duration).toArray
-      Arrays.sort(durations)
+      val durations = successfulTasksSet.toArray.map(taskInfos(_).duration)
```

Comment:
there are some other things you could do to make the original code more efficient:
a) instead of `.values.filter.map`, use a `foreach` to directly add into an `ArrayBuffer`.  That will avoid all the intermediate collections that scala would create otherwise.
b) store an approximate distribution of the runtimes, eg. using a tdigest.

aside: what  pain that there is no quick way to get the middle element of a TreeSet -- I couldn't find anything efficient in either the java or scala libs :(